### PR TITLE
Add an otelcol.processor.resourcedetection component

### PR DESCRIPTION
**Pull request by ptodev**
This is a draft PR to port over the [resource detection processor](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.87.0/processor/resourcedetectionprocessor/README.md). There are a few things which we need to agree on before the PR is taken out of "draft" stage. I will open comments on the specific lines where we need consensus.

### CODE REVIEW

File: component/otelcol/processor/resourcedetection/resourcedetection.go
Comment (ptodev) [inline component/otelcol/processor/resourcedetection/resourcedetection.go:64]: The HTTP settings in the Collector expose [HTTP client settings](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.87.0/processor/resourcedetectionprocessor/config.go#L37-L39) for two reasons:
* The timeout is used to cancel resource detections if they take too long. This use case has nothing to do with HTTP at all.
* The ec2 detection [uses the HTTP settings](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.87.0/processor/resourcedetectionprocessor/internal/aws/ec2/ec2.go#L101-L108) if they are provided. This feature is not documented in the OTel docs.

For now I would like to expose just the `timeout` parameter:
* This is because it would be super confusing to document `timeout` as both a timeout for HTTP and for ec2.
* I'll leave a TODO to implement it in the future.

I wish the OTel Collector worked differently:
* There should be a non-HTTP `timeout` settings for all detectors
* There should be HTTP settings specific to the ec2 detector. That way ec2 won't use `context`, and it will be possible to have different HTTP settings for different detectors.

Maybe it'll be work raising these issues with the upstream maintainers.

File: docs/sources/flow/reference/components/otelcol.processor.resourcedetection.md
Comment (ptodev) [inline docs/sources/flow/reference/components/otelcol.processor.resourcedetection.md:52]: I'd like to rename this section to "Configuration" and remove the table with blocks. In its place I could put a table with "root level" attributes and a table with "root level" blocks. This was already described in #5390

File: component/otelcol/processor/resourcedetection/internal/resource_attribute_config/resource_attribute_config.go
Comment (ptodev) [inline component/otelcol/processor/resourcedetection/internal/resource_attribute_config/resource_attribute_config.go:3]: Note that `otelcol.processor.resourcedetection` will work differently from the Collector's original processor:
* In the Collector, many resource attributes are enabled by default. But some resource attributes are not.
* In the Agent, all resource attributes will be disabled by default. Users will have to enable them explicitly.

There are a few reasons why I decided to do it this way:
* If the Agent sets various resource attributes by default, the Collector might print warnings which are confusing to the user. This is the `enabledSetByUser` parameter mentioned in the comment below.
* So that we don't have to waste time keeping the Agent's defaults in sync with the Collector's.

For example, this is a labda resource detection with all resource attributes enabled:
```
otelcol.processor.resourcedetection "default"  {
  detectors = ["lambda"]
  lambda {
    resource_attributes {
      aws.log.group.names { enabled = true }
      aws.log.stream.names { enabled = true }
      cloud.platform { enabled = true }
      cloud.provider { enabled = true }
      cloud.region { enabled = true }
    }
  }
  output {
    logs = [otelcol.exporter.otlp.default.input]
  }
}
```

If a user wants to disable a resource attribute, all they need to do is just not list it:
```
otelcol.processor.resourcedetection "default"  {
  detectors = ["lambda"]
  lambda {
    resource_attributes {
      aws.log.group.names { enabled = true }
    }
  }
  output {
    logs = [otelcol.exporter.otlp.default.input]
  }
}
```

They could also explicitly set it to `enabled = false`:
```
otelcol.processor.resourcedetection "default"  {
  detectors = ["lambda"]
  lambda {
    resource_attributes {
      aws.log.group.names { enabled = true }
      cloud.region { enabled = false }
    }
  }
  output {
    logs = [otelcol.exporter.otlp.default.input]
  }
}
```

File: docs/sources/flow/reference/components/otelcol.processor.resourcedetection.md
Comment (clayton-cornell) [reply docs/sources/flow/reference/components/otelcol.processor.resourcedetection.md:160]: @ptodev I know there's been no visible movement on PR #5381 but.... it's a good proposal, and I've been wanting to get started on implementing it across all Ref topics... just haven't got to it yet.

Also.. guidelines for docs layouts should be that.. guidelines. If following the guideline results in a mess, we should be flexible and adjust the approach. If there's a better way (even when that means adjusting existing docs) that makes things more clear and understandable, then I'd say go for it. In this case there's already a PR open suggesting the same change so this PR is not a change that exists in isolation.

File: component/otelcol/processor/resourcedetection/internal/resource_attribute_config/resource_attribute_config.go
Comment (mar4uk) [reply component/otelcol/processor/resourcedetection/internal/resource_attribute_config/resource_attribute_config.go:3]: I completely understand the approach you are suggesting and support it. The only concern I have is will this change prevent the `otelcol.processor.resourcedetection` component from becoming a distribution of otel collector resourcedetectionprocessor?

Comment (ptodev) [reply component/otelcol/processor/resourcedetection/internal/resource_attribute_config/resource_attribute_config.go:3]: @mar4uk I decided to backtrack on this, and instead I made the Agent component behave similarly to the Collector one. It will default certain resource attributes. I hope this makes things easier for users. ATM it defaults the same ones which the Collector defaults, but in the future it might diverge a bit. That should be ok.

File: docs/sources/flow/reference/components/otelcol.processor.resourcedetection.md
Comment (ptodev) [inline docs/sources/flow/reference/components/otelcol.processor.resourcedetection.md:88]: I need to check this:
https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.87.0/processor/resourcedetectionprocessor/README.md#gcp

File: docs/sources/flow/reference/components/otelcol.processor.resourcedetection.md
Comment (ptodev) [inline docs/sources/flow/reference/components/otelcol.processor.resourcedetection.md:163]: These `... > resource_attributes` headings have a small weight of `####` so that they don't appear in the ToC. This makes the ToC much cleaner.

File: docs/sources/flow/reference/components/otelcol.processor.resourcedetection.md
Comment (ptodev) [inline docs/sources/flow/reference/components/otelcol.processor.resourcedetection.md:396]: I need to check this too:
https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.87.0/processor/resourcedetectionprocessor/internal/consul/internal/metadata/generated_config.go#L28-L37

File: docs/sources/flow/reference/components/otelcol.processor.resourcedetection.md
Comment (ptodev) [reply docs/sources/flow/reference/components/otelcol.processor.resourcedetection.md]: I updated the PR to hide the `... > resource_attributes` headings in the ToC:
https://github.com/grafana/agent/pull/5764#discussion_r1447975010

File: component/otelcol/processor/resourcedetection/resourcedetection.go
Comment (rlankfo) [inline component/otelcol/processor/resourcedetection/resourcedetection.go:44]: I assume this should be `otelcol.processor.resourcedetection` here

**Comment by rlankfo** [comment]
Does this provide support for the env resource detector? e.g. setting resource attributes in `OTEL_RESOURCE_ATTRIBUTES` environment variable?

**Comment by ptodev** [comment]
> Does this provide support for the env resource detector? e.g. setting resource attributes in `OTEL_RESOURCE_ATTRIBUTES` environment variable?

@rlankfo - Yes. It is mentioned in the [docs](https://github.com/grafana/agent/pull/5764/files/4bb42f67714707d5c34b212e54c465417fb3e396#diff-ba78575ddc15f488c5e89aedeefea2e9ba4cfe5b2391a90edc3f972114c957aeR70-R73) and there is a [unit test](https://github.com/grafana/agent/pull/5764/files/4bb42f67714707d5c34b212e54c465417fb3e396#diff-5fa1e129bac5f26513f330cf291e03329447952192d0262f8b21b933cc720603R1457-R1484).

File: docs/sources/flow/reference/components/otelcol.processor.resourcedetection.md
Comment (mar4uk) [reply docs/sources/flow/reference/components/otelcol.processor.resourcedetection.md:88]: the resourcedetection readme needs a fix probably. I found that gce and gke were deprecated to use only one gcp detector https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/10347, ordering for gcp is not needed anymore if I understood correctly

File: component/otelcol/processor/resourcedetection/internal/openshift/config.go
Comment (mar4uk) [inline component/otelcol/processor/resourcedetection/internal/openshift/config.go:51]: nit: in all other config files `resourcedetectionprocessor/` prefix was added. Shall it be `resourcedetectionprocessor/openshift` instead of `openshift`. That's super minor

Comment (ptodev) [reply component/otelcol/processor/resourcedetection/internal/openshift/config.go:51]: Thanks for spotting this! I'll remove the `resourcedetectionprocessor/` because it feels unnecessary.

File: docs/sources/flow/reference/components/otelcol.processor.resourcedetection.md
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/otelcol.processor.resourcedetection.md:77]:
```suggestion
If you omit the [ec2][] block, the defaults specified in the [ec2][] block documentation are used.
```

File: docs/sources/flow/reference/components/otelcol.processor.resourcedetection.md
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/otelcol.processor.resourcedetection.md:80]:
```suggestion
For example, if you had `detectors = ["eks", "ec2"]` then `cloud.platform` will be `aws_eks` instead of `ec2`. 
```

File: docs/sources/flow/reference/components/otelcol.processor.resourcedetection.md
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/otelcol.processor.resourcedetection.md:82]:
```suggestion
The following order is recommended for AWS:
```

File: docs/sources/flow/reference/components/otelcol.processor.resourcedetection.md
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/otelcol.processor.resourcedetection.md:154]:
```suggestion
To fetch EC2 tags, the IAM role assigned to the EC2 instance must have a policy that includes the `ec2:DescribeTags` permission.
```

We should avoid using "note that". If it is actually a note, then we should use the note admonition shortcode.

The text "in order to" can almost always be simplified to "to".

File: docs/sources/flow/reference/components/otelcol.processor.resourcedetection.md
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/otelcol.processor.resourcedetection.md:496]:
```suggestion
One known issue happens when GKE workload identity is enabled. The GCE metadata endpoints won't be available, 
and the GKE resource detector won't be able to determine `host.name`. 
```

File: docs/sources/flow/reference/components/otelcol.processor.resourcedetection.md
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/otelcol.processor.resourcedetection.md:497]:
```suggestion
If this happens, you can set `host.name` from one of the following resources:
```

File: docs/sources/flow/reference/components/otelcol.processor.resourcedetection.md
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/otelcol.processor.resourcedetection.md:499]:
```suggestion
- Get the `node.name` through the downward API with the `env` detector.
- Get the Kubernetes node name from the Kubernetes API (with `k8s.io/client-go`).
```

What is a downward API.

File: docs/sources/flow/reference/components/otelcol.processor.resourcedetection.md
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/otelcol.processor.resourcedetection.md:471]: Are all these headings supposed to be initial caps? Google Compute Engine should be (it defines the GCE acronym) but is `Metadata` also supposed to be initial caps?

File: docs/sources/flow/reference/components/otelcol.processor.resourcedetection.md
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/otelcol.processor.resourcedetection.md:663]:
```suggestion
If TLS is not explicitly disabled and no `ca_file` is configured, `/var/run/secrets/kubernetes.io/serviceaccount/ca.crt` is used. 
```

File: docs/sources/flow/reference/components/otelcol.processor.resourcedetection.md
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/otelcol.processor.resourcedetection.md:664]:
```suggestion
The determination of the API address, `ca_file`, and the service token is skipped if they are set in the configuration.
```

File: docs/sources/flow/reference/components/otelcol.processor.resourcedetection.md
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/otelcol.processor.resourcedetection.md:715]:
```suggestion
`auth_type` can be set to one of the following:
```

File: docs/sources/flow/reference/components/otelcol.processor.resourcedetection.md
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/otelcol.processor.resourcedetection.md:803]:
```suggestion
There is no need to put in an `ec2 {}` River block.
```

File: docs/sources/flow/reference/components/otelcol.processor.resourcedetection.md
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/otelcol.processor.resourcedetection.md:882]:
```suggestion
You need to add this to your workload:
```

File: docs/sources/flow/reference/components/otelcol.processor.resourcedetection.md
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/otelcol.processor.resourcedetection.md:911]:
```suggestion
You need to add this to your workload:
```

File: docs/sources/flow/reference/components/otelcol.processor.resourcedetection.md
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/otelcol.processor.resourcedetection.md:934]:
```suggestion
Connecting some components may not be sensible, or components may require further configuration to make the 
```

**Review by clayton-cornell — COMMENTED** [review]
Some doc suggestions and questions

File: component/otelcol/processor/resourcedetection/internal/resource_attribute_config/resource_attribute_config.go
Comment (tpaschalis) [inline component/otelcol/processor/resourcedetection/internal/resource_attribute_config/resource_attribute_config.go:6]: I don't mean to block this PR, since this is already a big effort, but I'm curious, did we consider other possibilities to make this easier to configure and/or maintain in the future?

For example, one possibility was to consider making this a bool type alias, or an enum-like field

So instead of
```
cloud.account.id { enabled = true }
```

We either get
```
cloud.account.id = true
// or
enabled = ["cloud.account.id", "foo", "bar"]
```

Also, AFAICT in the upstream OTEL collector they use some kind of [YAML metadata](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/resourcedetectionprocessor/internal/aws/ec2/metadata.yaml) to generate all these config options, maybe we could use a similar approach?

**Review by tpaschalis — APPROVED** [review]
LGTM % some nits; I don't mean to block this PR as it's been a large effort, but I have a potential concern about the long-term maintainability of it and whether we could re-think how we configure these blocks and keep track of new options.

GIven that would be a big change in of itself, I'm fine with re-considering in a future PR.

File: docs/sources/flow/reference/components/otelcol.processor.resourcedetection.md
Comment (ptodev) [reply docs/sources/flow/reference/components/otelcol.processor.resourcedetection.md:49]: The detectors are ran by `otelcol.processor.resourcedetection`. For example, `ecs` and `consul` are detectors. I think `should be ran` is ok because it indicates that if a detector is no listed in `detectors`, it will not be ran at all.

Comment (clayton-cornell) [reply docs/sources/flow/reference/components/otelcol.processor.resourcedetection.md:49]: "ran" is past tense, so it's probably the wrong word here. "should" implies it's optional.. you should do it.. but you don't have to.

`An ordered list of named detectors used to detect resource information.`  This leaves it open to whichever process "owns" and runs the detectors.

Comment (ptodev) [reply docs/sources/flow/reference/components/otelcol.processor.resourcedetection.md:49]: Ok, thank you, I'll change it to that!

File: docs/sources/flow/reference/components/otelcol.processor.resourcedetection.md
Comment (ptodev) [reply docs/sources/flow/reference/components/otelcol.processor.resourcedetection.md:471]: There is no reason for `Metadata` that I can think of, so I will lower it.

File: docs/sources/flow/reference/components/otelcol.processor.resourcedetection.md
Comment (ptodev) [reply docs/sources/flow/reference/components/otelcol.processor.resourcedetection.md:499]: It is a way for a pod to get more information about the environment it is running in:
https://kubernetes.io/docs/concepts/workloads/pods/downward-api/

I'll change "downward API" to be a link to the k8s docs.

**Review by clayton-cornell — APPROVED** [review]
Docs look good to me

File: component/otelcol/processor/resourcedetection/resourcedetection.go
Comment (ptodev) [reply component/otelcol/processor/resourcedetection/resourcedetection.go]: Thanks for spotting this! I think the Collector code would throw an error, but it would be more correct for River to do it in the Validate() function. I updated the code to do just that, and added tests.

File: docs/sources/flow/reference/components/otelcol.processor.resourcedetection.md
Comment (ptodev) [reply docs/sources/flow/reference/components/otelcol.processor.resourcedetection.md]: I raised a PR to fix this upstream:
https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/30727

**Comment by mar4uk** [comment]
🎉

**Pull request merged**
